src/messages/validate.rs

```diff
@@ -0,0 +1,118 @@
+//! Validation rules specified in [RFC-001#rules]
+//!
+//! [RFC-001#rules]: https://github.com/ZcashFoundation/redjubjub/blob/main/rfcs/0001-messages.md#rules
+
+use super::constants::{
+    BASIC_FROST_SERIALIZATION, MAX_PROTOCOL_MESSAGE_LEN, MAX_SIGNER_PARTICIPANT_ID,
+};
+use super::*;
+
+use thiserror::Error;
+
+pub trait Validate {
+    fn validate(&self) -> Result<&Self, MsgErr>;
+}
+
+impl Validate for Message {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        match self.payload {
+            Payload::SharePackage(_) => {
+                if self.header.sender != ParticipantId::Dealer {
+                    return Err(MsgErr::SenderMustBeDealer);
+                }
+                if self.header.receiver <= ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
+            }
+            Payload::SigningCommitments(_) => {
+                if self.header.sender <= ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
+                    return Err(MsgErr::SenderMustBeSigner);
+                }
+                if self.header.receiver != ParticipantId::Aggregator {
+                    return Err(MsgErr::ReceiverMustBeAggergator);
+                }
+            }
+            Payload::SigningPackage(_) => {
+                if self.header.sender != ParticipantId::Aggregator {
+                    return Err(MsgErr::SenderMustBeAggregator);
+                }
+                if self.header.receiver <= ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
+            }
+            Payload::SignatureShare(_) => {
+                if self.header.sender <= ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
+                    return Err(MsgErr::SenderMustBeSigner);
+                }
+                if self.header.receiver != ParticipantId::Aggregator {
+                    return Err(MsgErr::ReceiverMustBeAggergator);
+                }
+            }
+            Payload::AggregateSignature(_) => {
+                if self.header.sender != ParticipantId::Aggregator {
+                    return Err(MsgErr::SenderMustBeAggregator);
+                }
+                if self.header.receiver <= ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
```

Comment:
The comparisons are the wrong way around - valid signer ids are less than or equal to the maximum.

Also the enum comparison rules depend on the order of the variants, so it's better to just do a match.
```suggestion
                if !matches!(self.header.receiver, ParticipantId::Signer(_)) {
                    return Err(MsgErr::ReceiverMustBeSigner);
                }
            }
            Payload::SigningCommitments(_) => {
                if !matches!(self.header.sender, ParticipantId::Signer(_)) {
                    return Err(MsgErr::SenderMustBeSigner);
                }
                if self.header.receiver != ParticipantId::Aggregator {
                    return Err(MsgErr::ReceiverMustBeAggergator);
                }
            }
            Payload::SigningPackage(_) => {
                if self.header.sender != ParticipantId::Aggregator {
                    return Err(MsgErr::SenderMustBeAggregator);
                }
                if !matches!(self.header.receiver, ParticipantId::Signer(_)) {
                    return Err(MsgErr::ReceiverMustBeSigner);
                }
            }
            Payload::SignatureShare(_) => {
                if self.header.sender > ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
                    return Err(MsgErr::SenderMustBeSigner);
                }
                if self.header.receiver != ParticipantId::Aggregator {
                    return Err(MsgErr::ReceiverMustBeAggergator);
                }
            }
            Payload::AggregateSignature(_) => {
                if self.header.sender != ParticipantId::Aggregator {
                    return Err(MsgErr::SenderMustBeAggregator);
                }
                if self.header.receiver > ParticipantId::Signer(MAX_SIGNER_PARTICIPANT_ID) {
                    return Err(MsgErr::ReceiverMustBeSigner);
                }
```